### ISSUE TITLE
Emoji can leave artifacts in right cell

### ISSUE THREAD

**Issue by Tyriar** [xtermjs/xterm.js#1502]
See https://github.com/Microsoft/vscode/issues/51385#issuecomment-395538985

![image](https://user-images.githubusercontent.com/2193314/41144377-0621383a-6afd-11e8-898c-78543fa66d45.png)

**Comment by jerch** [comment]
Two possible solutions come to my mind:
- mark cell as overdrawn by left cell and therefore redraw it when left cell gets new content
- scale bogus chars to fit in wcwidth * cells, no overdrawing at all

**Comment by jerch** [comment]
There is the `_isOverlapping` test in the code (https://github.com/xtermjs/xterm.js/blob/master/src/renderer/TextRenderLayer.ts#L84), but to me it seems it is only applied when an overlapping char gets rendered but not when it gets removed (commented out?) Maybe @bgw can have a look at it?

**Comment by Tyriar** [comment]
Yeah I think it used to, if a cell is overlapping it must be drawn at the same time as its neighbouring cell, and if it's removed the neighbouring cell needs to be redrawn as well.

**Comment by jerch** [comment]
Read the original issue - the example the OP gave with the aubergines looks like it is wcwidth related, where the system wcwidth gives 2 and xterm.js's 1. This would explain why the OP can move the cursor further back into the prompt. I dont have Ubuntu 18 here - can someone with Ubuntu 18 test the system wcwidth for `🍆`?

**Comment by bgw** [comment]
To elaborate on @Tyriar's point, we clear entire rows at a time: https://github.com/xtermjs/xterm.js/blob/fb63b3475b009b9f340703b28e0336c44030d82a/src/renderer/TextRenderLayer.ts#L213

There should be one `clearRect` or `fillRect` call (depending on alpha) per frame.

> can someone with Ubuntu 18 test the system wcwidth for 🍆?

I'm not on Ubuntu 18, but I'm on Debian Stable, which is a lot closer than Ubuntu 14. On my machine, wcwidth for 🍆 is 2:

![selection_064](https://user-images.githubusercontent.com/180404/41273064-c39b87b8-6dcc-11e8-8fc9-cc2fb5024783.png)

The program if anyone wants it for testing:

```c
#define _XOPEN_SOURCE
#include <stdio.h>
#include <stddef.h>
#include <wchar.h>
#include <locale.h>

int main() {
    wchar_t wc = L'\U0001F346';
    setlocale(LC_ALL, "en_US.UTF-8");
    wprintf(L"%d, %lc\n", wcwidth(wc), wc);
}
```

**Comment by Tyriar** [comment]
> we clear entire rows at a time

Oh I forgot about that 😅. I'm quite puzzled as to how this is happening then as the emoji would need to be drawn, but it looks like the class of bugs where the right is not "cleared" (which can't happen because we always clear and redraw the cells now).

**Comment by jerch** [comment]
@bgw Thanks for testing - yupp this is a bug in wcwidth, which returns 1 here. This is a case where the backend wcwidth and our wcwidth return different values for a specific char (see https://github.com/xtermjs/xterm.js/issues/1467#issuecomment-391941536 for the background). We can fix it by applying custom settings once #1470 is done.

**Comment by Tyriar** [comment]
I don't think this happens anymore.
